### ISSUE TITLE
idea: switch to sbt-proguard from sbt-assembly for building desktop jar

### ISSUE THREAD

**Issue by aginiewicz** [ajhager/libgdx-sbt-project.g8#40]
sbt-assembly "just works", but creates huge jar files, even for simple projects - it's like 16mb vs 1mb on android. I think that we could switch to sbt-proguard for it. IIRC there was comment on iOS ticket, that proguard will be useful there as well.

If we switch to sbt-proguard and do some tweaking to proguard config files handling, we will be able to share some configuration between desktop/ios/android - i.e. for desktop we could merge config from common and desktop, for android from common and android, and so on... if we put all scala related stuff in there, and all android related stuff into android project, it should be quite easy to just concatenate two config files.

Sadly, I believe we won't be able to use sbt-android bundled proguard, it seems to be too closely tied into android project and build process to use separately, unless I'm mistaken.

**Comment by ajhager** [comment]
I know I looked at this in the past and the proguard bundled with sbt-android was causing issues. I would love to bypass it and handle proguard ourselves. Proguard on desktop would be especially nice.

**Comment by aginiewicz** [comment]
I don't believe bypassing it will be doable without modification of sbt-android. I think that http://github.com/sbt/sbt-proguard released 6 months ago will do the trick, seems to be very easy and not doing anything fishy to configuration (just passing it in one-to-one manner).

**Comment by ajhager** [comment]
Let me know if you are working on this, otherwise I will see what I can come up with.

**Comment by aginiewicz** [comment]
Seems it won't be that easy after all:

```
[error] AttributeKey ID collisions detected for: 'proguard' (sbt.Task[scala.Option[java.io.File]], sbt.Task[scala.collection.Seq[java.io.File]]), 'proguard-configuration' (sbt.Task[scala.Option[java.io.File]], java.io.File)
```

seems that sbt-proguard and sbt-android are in conflict, because they both define keys 'proguard' and 'proguard-configuration'.

I will ask about it on sbt mailing list (to see if there is known workaround for case of conflicting plugins) or not, contact sbt-proguard/sbt-android devs so we can check if we are able to make those work.

But before this, I'm experimenting with manually creating task for desktop project and reusing proguard jar pulled by sbt-android. I will update this ticket with new findings.

**Comment by ajhager** [comment]
Thanks for working on this. Is there anyway to rename the conflicting keys on import?

For example:

import sbtandroid.AndroidPlugin.{proguard => proguardAndroid, proguard-configuration => proguard-configurationAndroid, _}

**Comment by aginiewicz** [comment]
I don't think so, because the problem is with key names like in

```
lazy val updateLibgdxKey = TaskKey[Unit]("update-gdx", "Updates libgdx")
```

anyway, I managed to run proguard from my own task and parse list of jars. It still needs some work, but there is output sbt run in which I've checked that I can access proguard:

```
> assembly
[info] preparing proguarded assembly
[info] ProGuard, version 4.8
[info] Usage: java proguard.ProGuard [options ...]
[success] Total time: 1 s, completed 2013-09-22 16:59:56
```

which means we will have solution that is more on-par with proguard used by sbt-android.

**Comment by ajhager** [comment]
Ah yes, good point. I know there are some changes in 0.13 that should help with conflicting keys, but I have a feeling it is going to be a while before we can upgrade.

**Comment by aginiewicz** [comment]
https://github.com/ajhager/libgdx-sbt-project.g8/pull/50 can you try it out? It worked for my test project :)

**Comment by ajhager** [comment]
Fixed by #61
